packages/backend-for-frontend/src/services/toolService.ts

```diff
@@ -0,0 +1,454 @@
+/* eslint-disable @typescript-eslint/explicit-function-return-type */
+
+import { isAxiosError } from "axios";
+import {
+  ApiKey,
+  ClientAssertion,
+  ConsumerKey,
+  FailedValidation,
+  SuccessfulValidation,
+  validateClientKindAndPlatformState,
+  validateRequestParameters,
+  verifyClientAssertion,
+  verifyClientAssertionSignature,
+} from "pagopa-interop-client-assertion-validation";
+import {
+  AgreementId,
+  ApiError,
+  ClientId,
+  EServiceId,
+  ItemState,
+  PurposeId,
+  TenantId,
+  unsafeBrandId,
+} from "pagopa-interop-models";
+import { WithLogger } from "pagopa-interop-commons";
+import {
+  agreementApi,
+  authorizationApi,
+  bffApi,
+  catalogApi,
+  purposeApi,
+} from "pagopa-interop-api-clients";
+import { BffAppContext } from "../utilities/context.js";
+import {
+  activeAgreementByEserviceAndConsumerNotFound,
+  agreementDescriptorNotFound,
+  agreementNotFound,
+  cannotGetKeyWithClient,
+  clientAssertionPublicKeyNotFound,
+  ErrorCodes,
+  eserviceDescriptorNotFound,
+  missingActivePurposeVersion,
+  organizationNotAllowed,
+  purposeIdNotFoundInClientAssertion,
+} from "../model/errors.js";
+import {
+  PagoPAInteropBeClients,
+  AgreementProcessClient,
+  CatalogProcessClient,
+} from "../clients/clientsProvider.js";
+import { getAllAgreements, getLatestAgreement } from "./agreementService.js";
+
+export function toolsServiceBuilder(clients: PagoPAInteropBeClients) {
+  return {
+    async validateTokenGeneration(
+      clientId: string | undefined,
+      clientAssertion: string,
+      clientAssertionType: string,
+      grantType: string,
+      ctx: WithLogger<BffAppContext>
+    ): Promise<bffApi.TokenGenerationValidationResult> {
+      ctx.logger.info(`Validating token generation for client ${clientId}`);
+
+      const { errors: parametersErrors } = validateRequestParameters({
+        client_assertion: clientAssertion,
+        client_assertion_type: clientAssertionType,
+        grant_type: grantType,
+        client_id: clientId,
+      });
+
+      const { data: jwt, errors: clientAssertionErrors } =
+        verifyClientAssertion(clientAssertion, clientId);
+      if (parametersErrors || clientAssertionErrors) {
```

Comment:
```suggestion

      if (parametersErrors || clientAssertionErrors) {
```
Very minor point: I think it's preferable to make this check standout on its own